compactor/periodic.go

```diff
@@ -38,60 +38,60 @@ type Periodic struct {
 	ctx    context.Context
 	cancel context.CancelFunc
 
-	mu     sync.Mutex
+	// mu protects paused
+	mu     sync.RWMutex
 	paused bool
 }
 
 // NewPeriodic creates a new instance of Periodic compactor that purges
-// the log older than h hours.
-func NewPeriodic(h int, rg RevGetter, c Compactable) *Periodic {
+// the log older than h Duration.
+func NewPeriodic(h time.Duration, rg RevGetter, c Compactable) *Periodic {
 	return &Periodic{
-		clock:        clockwork.NewRealClock(),
-		periodInHour: h,
-		rg:           rg,
-		c:            c,
+		clock:  clockwork.NewRealClock(),
+		period: h,
+		rg:     rg,
+		c:      c,
 	}
 }
 
+// periodDivisor divides Periodic.period in into checkCompactInterval duration
+const periodDivisor = 10
+
 func (t *Periodic) Run() {
 	t.ctx, t.cancel = context.WithCancel(context.Background())
 	t.revs = make([]int64, 0)
 	clock := t.clock
-
+	checkCompactInterval := t.period / time.Duration(periodDivisor)
 	go func() {
 		last := clock.Now()
 		for {
 			t.revs = append(t.revs, t.rg.Rev())
 			select {
 			case <-t.ctx.Done():
 				return
-			case <-clock.After(checkCompactionInterval):
+			case <-clock.After(checkCompactInterval):
 				t.mu.Lock()
 				p := t.paused
 				t.mu.Unlock()
 				if p {
 					continue
 				}
 			}
-
-			if clock.Now().Sub(last) < executeCompactionInterval {
+			if clock.Now().Sub(last) < t.period {
 				continue
 			}
-
-			rev, remaining := t.getRev(t.periodInHour)
+			rev, remaining := t.getRev()
 			if rev < 0 {
 				continue
 			}
-
-			plog.Noticef("Starting auto-compaction at revision %d (retention: %d hours)", rev, t.periodInHour)
+			plog.Noticef("Starting auto-compaction at revision %d (retention: %v)", rev, t.period)
 			_, err := t.c.Compact(t.ctx, &pb.CompactionRequest{Revision: rev})
 			if err == nil || err == mvcc.ErrCompacted {
 				t.revs = remaining
-				last = clock.Now()
```

Comment:
why we delete this line?

Reply:
good a catch. I don't think this is meant to be deleted.

Reply:
Actually, I dig into on why `last = clock.Now()` is deleted again. It is explained:

`I change the compaction frequency to every checkCompactionInterval when T > retentionPeriod.`
https://github.com/coreos/etcd/pull/8563#issuecomment-332364722

So the current behavior is not a bug but an behavior changed. I guess the question to ask is whether this intended behavior is correct or not.